contrib/verify-commits/verify-commits.py

```diff
@@ -0,0 +1,136 @@
+from sys import stdin,stdout,stderr
+import subprocess
+from subprocess import PIPE
+import hashlib
+
+GIT = os.getenv('GIT','git')
+
+def tree_sha512sum(commit='HEAD'):
+    # request metadata for entire tree, recursively
+    files = []
+    blob_by_name = {}
+    for line in subprocess.check_output([GIT, 'ls-tree', '--full-tree', '-r', commit]).splitlines():
+        name_sep = line.index(b'\t')
+        metadata = line[:name_sep].split() # perms, 'blob', blobid
+        assert(metadata[1] == b'blob')
+        name = line[name_sep+1:]
+        files.append(name)
+        blob_by_name[name] = metadata[2]
+
+    files.sort()
+    # open connection to git-cat-file in batch mode to request data for all blobs
+    # this is much faster than launching it per file
+    p = subprocess.Popen([GIT, 'cat-file', '--batch'], stdout=PIPE, stdin=PIPE)
+    overall = hashlib.sha512()
+    for f in files:
+        blob = blob_by_name[f]
+        # request blob
+        p.stdin.write(blob + b'\n')
+        p.stdin.flush()
+        # read header: blob, "blob", size
+        reply = p.stdout.readline().split()
+        assert(reply[0] == blob and reply[1] == b'blob')
+        size = int(reply[2])
+        # hash the blob data
+        intern = hashlib.sha512()
+        ptr = 0
+        while ptr < size:
+            bs = min(65536, size - ptr)
+            piece = p.stdout.read(bs)
+            if len(piece) == bs:
+                intern.update(piece)
+            else:
+                raise IOError('Premature EOF reading git cat-file output')
+            ptr += bs
+        dig = intern.hexdigest()
+        assert(p.stdout.read(1) == b'\n') # ignore LF that follows blob data
+        # update overall hash with file hash
+        overall.update(dig.encode("utf-8"))
+        overall.update("  ".encode("utf-8"))
+        overall.update(f)
+        overall.update("\n".encode("utf-8"))
+    p.stdin.close()
+    if p.wait():
+        raise IOError('Non-zero return value executing git cat-file')
+    return overall.hexdigest()
+
+def main():
+    dir = os.path.dirname(os.path.abspath(__file__))
+    print("Using verify-commits data from " + dir)
+
+    VERIFIED_ROOT = open(dir + "/trusted-git-root", "r").read().splitlines()[0]
+    VERIFIED_SHA512_ROOT = open(dir + "/trusted-sha512-root-commit", "r").read().splitlines()[0]
+    REVSIG_ALLOWED = open(dir + "/allow-revsig-commits", "r").read().splitlines()
+    DIRTY_ALLOWED = open(dir + "/allow-dirty-commits", "r").read().splitlines()
+
+    CURRENT_COMMIT = "HEAD" if len(sys.argv) == 1 else sys.argv[1]
+
+    if ' ' in CURRENT_COMMIT:
+        print("Commit must not contain spaces?", file=sys.stderr)
+        exit(1)
+
+    VERIFY_TREE = True if len(sys.argv) >= 3 and sys.argv[2] == "--tree-check" else False
+    NO_SHA1 = True
+    PREV_COMMIT = ""
+    INITIAL_COMMIT = CURRENT_COMMIT
+
+    BRANCH = subprocess.check_output([GIT,'rev-parse','--abbrev-ref','HEAD'], universal_newlines=True).splitlines()[0]
```

Comment:
Shouldn't this be `INITIAL_COMMIT` instead of `HEAD`?

Reply:
It's a direct transcription from the shell script at least:
```bash
BRANCH="$(git rev-parse --abbrev-ref HEAD)"
```

Reply:
No, it is a string 'HEAD', it would return a branch name. If you put a commit id, it returns nothing.

Reply:
Since @ken2812221 reports, that this doesn't work either way, we might just default this to `"master"`, unless overridden by an option.

Reply:
> 'HEAD', it would return a branch name. If you put a commit id, it returns nothing.

It always returns a commit for me.

Reply:
Well, the git version might be different. I'm using git 2.1.4

Reply:
Oh yeah, that's it. To me this seems like a thing to fix (The script shouldn't depends on what version of git you are using)

Also, I still think that the correct thing to use here is `INITIAL_COMMIT`. I.e. `git log --format="%H" -1 "${INITIAL_COMMIT}"`. Otherwise a comment might be appropriate to say why `HEAD` is used here.

Reply:
I use it just because the original shell script, so I change it to `git show -s --format="%H" ${INITIAL_COMMIT}` now, and looks like we don't need to change the travis script.